File: backend/app/services/user_service.py
"""用户服务：个人资料、头像、改密、注销。"""

from __future__ import annotations

import io
import secrets
import time

from fastapi import UploadFile
from PIL import Image
from sqlalchemy.orm import Session

from app.config import get_storage_root, settings
from app.core.security import hash_password, rsa_decrypt, validate_password, verify_password
from app.models.file_metadata import FileMetadata
from app.models.users import User
from app.schemas.common import AppError, ErrorCode
from app.schemas.user import MeOut, MeUpdateIn, OtherUserOut, PasswordChangeIn
from app.storage import filekit


def avatar_url_for(user: User) -> str:
    if user.avatar_path:
        return f"/api/v1/avatars/{user.id}"
    return "/api/v1/avatars/default"


def user_to_me_out(user: User) -> dict:
    return MeOut(
        id=user.id,
        email=user.email,
        nickname=user.nickname,
        signature=user.signature,
        avatar_url=avatar_url_for(user),
        role=user.role,
        status=user.status,
        can_invite=user.can_invite,
        created_at=user.created_at,
    ).model_dump(mode="json")


def get_me(db: Session, user: User) -> dict:
    return user_to_me_out(user)


def update_me(db: Session, user: User, data: MeUpdateIn) -> User:
    if data.nickname is not None:
        user.nickname = data.nickname
    if data.signature is not None:
        user.signature = data.signature
    db.commit()
    db.refresh(user)
    return user


def change_password(db: Session, user: User, data: PasswordChangeIn) -> None:
    from app.services.rsa_service import get_or_create_rsa_key

    key = get_or_create_rsa_key(db)

    try:
        old_plain = rsa_decrypt(key.private_key_pem, data.old_password)
    except ValueError:
        raise AppError(ErrorCode.RSA_DECRYPT_FAILED, "旧密码解密失败", 400) from None

    if not verify_password(old_plain, user.password_hash):
        raise AppError(ErrorCode.INVALID_CREDENTIALS, "旧密码不正确", 400)

    try:
        new_plain = rsa_decrypt(key.private_key_pem, data.new_password)
    except ValueError:
        raise AppError(ErrorCode.RSA_DECRYPT_FAILED, "新密码解密失败", 400) from None

    pw_errors = validate_password(new_plain)
    if pw_errors:
        raise AppError(ErrorCode.PASSWORD_TOO_WEAK, "密码强度不足", 400, {"errors": pw_errors})

    user.password_hash = hash_password(new_plain)
    db.commit()


def upload_avatar(db: Session, user: User, file: UploadFile) -> User:
    """上传头像：校验 -> 缩略图 -> 存 file_metadata。"""
    if not file.filename:
        raise AppError(ErrorCode.VALIDATION_ERROR, "未提供文件", 400)

    file.file.seek(0)
    content = file.file.read()
    max_size = settings.MEDIA_IMAGE_MAX_MB * 1024 * 1024
    if len(content) > max_size:
        raise AppError(ErrorCode.FILE_TOO_LARGE, "文件过大", 413)

    # 用 filekit 做真实格式校验 + 危险文件拒绝
    kind = filekit.detect_kind(content)
    if kind != "image":
        raise AppError(ErrorCode.UNSUPPORTED_MEDIA, "头像必须是图片", 400)
    filekit.check_size(kind, len(content))
    fmt, _ = filekit.validate_image(content)
    if fmt == "jpg":
        fmt = "jpeg"

    try:
        Image.open(io.BytesIO(content)).verify()
        img = Image.open(io.BytesIO(content))
    except Exception:
        raise AppError(ErrorCode.UNSUPPORTED_MEDIA, "图片文件无效", 400) from None

    storage_root = get_storage_root()
    avatars_dir = storage_root / "avatars"
    avatars_dir.mkdir(parents=True, exist_ok=True)

    ts = int(time.time())
    rand = secrets.token_hex(2)
    thumb_name = f"{user.id}_{ts}_{rand}_thumb.webp"
    thumb_path = avatars_dir / thumb_name

    img.thumbnail((settings.THUMB_SIZE, settings.THUMB_SIZE))
    if img.mode in ("RGBA", "P"):
        img = img.convert("RGB")  # type: ignore[assignment]
    img.save(thumb_path, format="WEBP", quality=settings.WEBP_THUMB_QUALITY)

    rel_path = f"avatars/{thumb_name}"
    user.avatar_path = rel_path

    meta = FileMetadata(
        storage_path=rel_path,
        original_name=file.filename or "avatar",
        filename=thumb_name,
        size=thumb_path.stat().st_size,
        mime="image/webp",
        format="webp",
        kind="avatar",
        owner_id=user.id,
    )
    db.add(meta)
    db.commit()
    db.refresh(user)
    return user


def deactivate(db: Session, user: User) -> None:
    if user.role == "admin":
        raise AppError(ErrorCode.FORBIDDEN, "管理员账号不可注销", 403)
    user.status = "deactivated"
    db.commit()


def get_other_user(db: Session, viewer_id: int, user_id: int) -> dict:
    from app.services.friend_service import get_friendship_status

    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise AppError(ErrorCode.NOT_FOUND, "用户不存在", 404)

    is_deactivated = user.status == "deactivated"
    avatar_url = "/api/v1/avatars/default" if is_deactivated else avatar_url_for(user)
    friendship_status = get_friendship_status(db, viewer_id, user_id)

    return OtherUserOut(
        id=user.id,
        nickname=user.nickname,
        signature=user.signature,
        avatar_url=avatar_url,
        is_deactivated=is_deactivated,
        created_at=user.created_at,
        friendship_status=friendship_status,
    ).model_dump(mode="json")

File: backend/internal/service/user_service.go
package service

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"

	"backend/internal/config"
	"backend/internal/model"
	"backend/internal/repository"
	"backend/internal/util"
)

type UserService struct {
	cfg   *config.Config
	repos *repository.Repositories
}

func NewUserService(cfg *config.Config, repos *repository.Repositories) *UserService {
	return &UserService{cfg: cfg, repos: repos}
}

func (s *UserService) AvatarURLFor(user *model.User) string {
	if user != nil && user.AvatarPath != nil && *user.AvatarPath != "" {
		return fmt.Sprintf("/api/v1/avatars/%d", user.ID)
	}
	return "/api/v1/avatars/default"
}

func (s *UserService) UserToMeOut(user *model.User) model.MeOut {
	return model.MeOut{
		ID:        user.ID,
		Email:     user.Email,
		Nickname:  user.Nickname,
		Signature: user.Signature,
		AvatarURL: s.AvatarURLFor(user),
		Role:      user.Role,
		Status:    user.Status,
		CanInvite: user.CanInvite,
		CreatedAt: user.CreatedAt,
	}
}

func (s *UserService) GetMe(ctx context.Context, user *model.User) (model.MeOut, error) {
	// Re-fetch to get fresh data
	fresh, err := s.repos.User.GetByID(ctx, user.ID)
	if err != nil {
		return model.MeOut{}, err
	}
	if fresh != nil {
		user = fresh
	}
	return s.UserToMeOut(user), nil
}

func (s *UserService) UpdateMe(ctx context.Context, user *model.User, data model.MeUpdateIn) (*model.User, error) {
	if err := s.repos.User.UpdateProfile(ctx, user.ID, data.Nickname, data.Signature); err != nil {
		return nil, err
	}
	fresh, err := s.repos.User.GetByID(ctx, user.ID)
	if err != nil {
		return nil, err
	}
	return fresh, nil
}

func (s *UserService) ChangePassword(ctx context.Context, user *model.User, data model.PasswordChangeIn) error {
	rsaKey, err := s.repos.RSA.GetFirst(ctx)
	if err != nil || rsaKey == nil {
		return model.NewAppError(model.ErrInternal, "RSA密钥未就绪", 500)
	}

	plainOld, err := util.RSADecrypt(rsaKey.PrivateKeyPEM, data.OldPassword)
	if err != nil {
		return model.NewAppError(model.ErrRSADecryptFailed, "旧密码解密失败", 400)
	}

	plainNew, err := util.RSADecrypt(rsaKey.PrivateKeyPEM, data.NewPassword)
	if err != nil {
		return model.NewAppError(model.ErrRSADecryptFailed, "新密码解密失败", 400)
	}

	if !util.VerifyPassword(plainOld, user.PasswordHash) {
		return model.NewAppError(model.ErrInvalidCredentials, "旧密码错误", 401)
	}

	pwErrors := util.ValidatePassword(plainNew)
	if len(pwErrors) > 0 {
		return model.NewAppError(model.ErrPasswordTooWeak, "新密码强度不足", 400, map[string]any{"errors": pwErrors})
	}

	newHash, err := util.HashPassword(plainNew)
	if err != nil {
		return model.NewAppError(model.ErrInternal, "新密码哈希失败", 500)
	}

	return s.repos.User.UpdatePasswordAndBumpToken(ctx, user.ID, newHash)
}

func (s *UserService) UploadAvatar(ctx context.Context, user *model.User, originalName string, content []byte) (*model.User, error) {
	maxBytes := s.cfg.MediaImageMaxMB * 1024 * 1024
	if len(content) > maxBytes {
		return nil, model.NewAppError(model.ErrFileTooLarge, "文件过大", 413)
	}

	kind, _, _, err := util.DetectKind(content)
	if err != nil || kind != "image" {
		return nil, model.NewAppError(model.ErrUnsupportedMedia, "不是有效的图片文件", 400)
	}

	avatarsDir, err := util.GetAvatarsDir(s.cfg.StorageRoot)
	if err != nil {
		return nil, model.NewAppError(model.ErrInternal, "创建头像目录失败", 500)
	}

	// Temporary file to run ffmpeg
	tmpFile, err := os.CreateTemp(avatarsDir, "avatar_upload_*.tmp")
	if err != nil {
		return nil, err
	}
	defer os.Remove(tmpFile.Name())

	if _, err := tmpFile.Write(content); err != nil {
		tmpFile.Close()
		return nil, err
	}
	tmpFile.Close()

	// Convert avatar to WebP thumbnail (max 400px)
	filename := util.GenerateFilename("webp")
	avatarAbsPath := filepath.Join(avatarsDir, filename)

	cmd := exec.Command("ffmpeg", "-y", "-i", tmpFile.Name(),
		"-vf", fmt.Sprintf("scale='min(%d,iw)':-2", s.cfg.ThumbSize),
		"-quality", fmt.Sprintf("%d", s.cfg.WebpThumbQuality),
		avatarAbsPath)
	if out, err := cmd.CombinedOutput(); err != nil {
		// If ffmpeg fails, fallback to direct save
		_ = out
		if err := os.WriteFile(avatarAbsPath, content, 0644); err != nil {
			return nil, model.NewAppError(model.ErrInternal, "保存头像文件失败", 500)
		}
	}

	relPath, err := filepath.Rel(s.cfg.StorageRoot, avatarAbsPath)
	if err != nil {
		relPath = filepath.Join("avatars", filename)
	}

	if err := s.repos.User.UpdateAvatar(ctx, user.ID, relPath); err != nil {
		return nil, err
	}

	fi, _ := os.Stat(avatarAbsPath)
	fileSize := int(fi.Size())

	_ = s.repos.Media.CreateFileMetadata(ctx, &model.FileMetadata{
		StoragePath:  relPath,
		OriginalName: originalName,
		Filename:     filename,
		Size:         fileSize,
		Mime:         "image/webp",
		Format:       "webp",
		Kind:         "avatar",
		OwnerID:      &user.ID,
	})

	user.AvatarPath = &relPath
	return user, nil
}

func (s *UserService) Deactivate(ctx context.Context, user *model.User) error {
	return s.repos.User.Deactivate(ctx, user.ID)
}

func (s *UserService) GetOtherUser(ctx context.Context, viewerID, targetUserID int) (*model.OtherUserOut, error) {
	target, err := s.repos.User.GetByID(ctx, targetUserID)
	if err != nil {
		return nil, err
	}
	if target == nil {
		return nil, model.NewAppError(model.ErrUserNotFound, "用户不存在", 404)
	}

	friendshipStatus := "none"
	if viewerID == targetUserID {
		friendshipStatus = "self"
	} else {
		areFriends, err := s.repos.Friend.AreFriends(ctx, viewerID, targetUserID)
		if err != nil {
			return nil, err
		}
		if areFriends {
			friendshipStatus = "friends"
		} else {
			f, err := s.repos.Friend.GetFriendship(ctx, viewerID, targetUserID)
			if err != nil {
				return nil, err
			}
			if f != nil && f.Status == "pending" {
				if f.RequesterID == viewerID {
					friendshipStatus = "pending_sent"
				} else {
					friendshipStatus = "pending_received"
				}
			}
		}
	}

	isDeactivated := target.Status == "deactivated"
	nickname := target.Nickname
	avatarURL := s.AvatarURLFor(target)
	if isDeactivated {
		nickname = "已注销"
		avatarURL = "/api/v1/avatars/default"
	}

	var isBot bool
	var personaBrief *string
	if target.Role == "bot" {
		bot, _ := s.repos.Bot.GetByUserID(ctx, target.ID)
		if bot != nil {
			isBot = true
			runes := []rune(bot.Persona)
			brief := string(runes)
			if len(runes) > 80 {
				brief = string(runes[:80])
			}
			personaBrief = &brief
		}
	}

	return &model.OtherUserOut{
		ID:               target.ID,
		Nickname:         nickname,
		Signature:        target.Signature,
		AvatarURL:        avatarURL,
		IsDeactivated:    isDeactivated,
		CreatedAt:        target.CreatedAt,
		FriendshipStatus: friendshipStatus,
		IsBot:            isBot,
		PersonaBrief:     personaBrief,
	}, nil
}
